src/corefx/System.Globalization.Native/casing.cpp

```diff
@@ -0,0 +1,120 @@
+//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+#include <assert.h>
+#include <stdint.h>
+#include <unicode/uchar.h>
+#include <unicode/utf16.h>
+
+/*
+Function:
+ToUpperSimple
+*/
+extern "C" void ToUpperSimple(const UChar* lpSrc, int32_t cwSrcLength, UChar* lpDst, int32_t cwDstLength)
+{
+    int32_t srcIdx = 0;
+    int32_t dstIdx = 0;
+
+    UBool isError = FALSE;
+
+    while (srcIdx < cwSrcLength)
+    {
+        UChar32 srcCodepoint;
+        UChar32 dstCodepoint;
+
+        U16_NEXT(lpSrc, srcIdx, cwSrcLength, srcCodepoint);
+        dstCodepoint = u_toupper(srcCodepoint);
+
+        U16_APPEND(lpDst, dstIdx, cwDstLength, dstCodepoint, isError);
+
+        // Ensure that we wrote the data and the source codepoint when encoded in UTF16 is the same
+        // number of code units as the cased codepoint.
+        assert(isError == FALSE && srcIdx == dstIdx);
+    }
+}
+
+/*
+Function:
+ToLowerSimple
+*/
+extern "C" void ToLowerSimple(const UChar* lpSrc, int32_t cwSrcLength, UChar* lpDst, int32_t cwDstLength)
```

Comment:
> Do you think this is premature?

You have a better sense than in in these scenarios for how hot the various code paths will be. If you think this is an area we will need such optimizations, go for it.  If you're not sure, we can do the simpler / less code duplication thing to start, and change it if/when we measure an impact (but we should definitely measure when we get to that point).  I'm fine with either.